### ISSUE TITLE
Implications of semaphore mechanism for SPARQL Update

### ISSUE THREAD

**Issue by kjetilk** [solid/specification#322]
While writing up a text around the semaphore mechanism from [the RWW Design Issue](https://www.w3.org/DesignIssues/ReadWriteLinkedData.html), also discussed in #125 , https://github.com/solid/solid-spec/pull/193 , https://github.com/solid-archive/query-panel/issues/3 and others, I found that I have not yet fully understood the problem. 

Is it also the idea that the same mechanism can be used in a situation where there isn't actual concurrency? I could see a case for that, say that one client `GET`s a certain resource, looks at it, and then wants to change something using a `PATCH` but meanwhile another client has changed the same? This is more the archetypal case for conditional requests, so I didn't think of it as a valid situation for the semaphore mechanism, but I guess a case for that this is actually a simpler mechanism could be made, as maintaining validators are more work for servers, and since it is a validator for the entire document, it scales badly if many people are co-editing a large document. 

This issue has usually been discussed in terms of _atomicity_, including [rdflib's documentation](https://github.com/linkeddata/rdflib.js/blob/8b405a1cef8a576cad0b1d4233201bd40e45b959/Documentation/webapp-intro.html#L473). I have not quite managed to understand that angle in light of the SPARQL Update spec, [which says](https://www.w3.org/TR/sparql11-update/#deleteInsert):

> If any solution produces a triple containing an unbound variable or an illegal RDF construct, such as a literal in a subject or predicate position, then that triple is not included when processing the operation: INSERT will not instantiate new data in the output graph, and DELETE will not remove anything. 

In our case, it means that if the `WHERE` clause doesn't match anything, the variable will be unbound and so nothing will be deleted. In other words, it sounds like we have standard behavior from the _query language_. 

However, it is true that no error is raised by standard SPARQL, as a measure to not leak information. Thus, we must expect that standard SPARQL implementations will need possibly substantial modifications to accommodate for this case now.

To accommodate for this situation on the Solid protocol, I wonder if it would suffice to say something like 

"If any solution produces a triple containing an unbound variable or an illegal RDF construct, then the server MUST abort any modifications and respond with a `409` status code."

That seems like a relatively minor change compared to standard SPARQL. This wouldn't be a violation of the language, AFAICS, it is a protocol thing.

I have only seen the semaphore mechanism mention in connection to the `DELETE INSERT WHERE` case, does that mean we only have to consider it in that context? 

In the case where there is a pure delete operation, you don't actually care if someone has deleted the data before you, it is gone, and then all is fine, right?

It seems to me like the `DELETE DATA`/`INSERT DATA` case is a bigger problem than the `DELETE INSERT WHERE` case, because in SPARQL, they are considered separate operations, even if executed in a single HTTP request. This would lead to that both `INSERT DATA`s are committed, isolation or not.

The easy way out of that is that developers should always use `DELETE INSERT WHERE` if they need semaphore behavior. That would, as far as I can see, be compatible with NSS. 

The alternative is to say that if `DELETE DATA` deletes nothing, then the server must return `409`, which is an actual spec violation. It may not be very different practically, I suppose, but still a spec violation.

**Comment by RubenVerborgh** [comment]
> I found that I have not yet fully understood the problem.

Here is a typical case that people had in mind for this with Databrowser.

- My color of a meeting participation record is blue
- I want to change that color to red
- So hence a `PATCH` request is created to delete the blue and add to red.
- However, if there were zero matches for the blue, then that means the color was changed already. So abort.
- If there was more than one match for the color, then the state is broken, so abort.

The actual code for this can be traced back all the way to rdflib.js, as can be seen here: https://github.com/linkeddata/rdflib.js/pull/298/files

> I asked in [#125 (comment)](https://github.com/solid/specification/issues/125#issuecomment-944345692) whether we should support several operations in one request, and @RubenVerborgh 's response was "not yet".

Yes; unfortunately I have now seen that Databrowser and NSS interact as follows:

```SPARQL
DELETE DATA {
  <https://ruben2021.solidcommunity.net/profile/card#me> <http://www.w3.org/2006/vcard/ns#fn> "Kjetil" .
 }
 ; INSERT DATA { <https://ruben2021.solidcommunity.net/profile/card#me> <http://www.w3.org/2006/vcard/ns#fn> "Ruben" .
 }
```

So note how this is done with two queries, without a `WHERE` clause, and without matching variables.

And unfortunately, it will follow the _same_ logic: if the deletion fails because there are no or multiple matches, then neither the insertion nor the deletion happen.

**Comment by kjetilk** [comment]
Right, I suspected as much for the  `DELETE DATA` `INSERT DATA` case, but I can't think of any way to reconcile the multiple match case with SPARQL. I need to think further about that.

**Comment by kjetilk** [comment]
I might also add that exactly that kind of query 
```sparql
DELETE DATA {
  <https://ruben2021.solidcommunity.net/profile/card#me> <http://www.w3.org/2006/vcard/ns#fn> "Kjetil" .
 }
 ; INSERT DATA { <https://ruben2021.solidcommunity.net/profile/card#me> <http://www.w3.org/2006/vcard/ns#fn> "Ruben" .
 }
```
was the reason I originally proposed to have just `DELETE DATA; INSERT DATA` as the subset. If you don't have any variables, it is an awful lot easier to have just one match :-) If all those queries that require only one match could be rewritten as this one, that would also make life much easier.
